### PR TITLE
Update firmware version management

### PR DESCRIPTION
**Pull request by jeromecoutant**
To make version check easier, version is now an integer.

DISCO_F413ZH version : 3522
IOT board version can be updated.

md file has been updated.

### CODE REVIEW

File: ISM43362Interface.cpp [hunk 1]
```diff
@@ -51,23 +44,21 @@ ISM43362Interface::ISM43362Interface(bool debug)
...
     }
-    debug_if(_ism_debug, "ISM43362Interface: read_version = [%s]\r\n", read_version);
-
-    if ((strcmp(read_version, supported_fw_versions[0]) == 0) || (strcmp(read_version, supported_fw_versions[1]) == 0)) {
-        debug_if(_ism_debug, "ISM43362Interface: firmware version is OK\r\n");
-    } else {
-        debug_if(_ism_debug, "ISM43362Interface: WARNING this firmware version has not been tested !\r\n");
+    debug_if(_ism_debug, "ISM43362Interface: read_version = %lu\r\n", _FwVersion);
+#if TARGET_DISCO_L475VG_IOT01A
+    if (_FwVersion < 3525) {
```
Comment (bcostm) [inline ISM43362Interface.cpp]: Maybe use a #define for this number ? Like LATEST_FW_VERSION_NUMBER ?

Comment (jeromecoutant) [reply ISM43362Interface.cpp]: I like it, thx

File: ISM43362/ISM43362.cpp [hunk 2]
```diff
@@ -60,6 +61,10 @@ extern "C" int32_t ParseNumber(char *ptr, uint8_t *cnt)
         ptr++;
         i++;
     }
+    if (*ptr == 'C') {  /* input string for version is C3.x.x.x */
```
Comment (bcostm) [inline ISM43362/ISM43362.cpp]: Maybe we can add this test in the while loop below instead ?

Comment (LMESTM) [reply ISM43362/ISM43362.cpp]: What if an X is 10 or more ?

File: ISM43362Interface.cpp [hunk 3]
```diff
@@ -51,23 +46,21 @@ ISM43362Interface::ISM43362Interface(bool debug)
...
-
-    if ((strcmp(read_version, supported_fw_versions[0]) == 0) || (strcmp(read_version, supported_fw_versions[1]) == 0)) {
-        debug_if(_ism_debug, "ISM43362Interface: firmware version is OK\r\n");
-    } else {
-        debug_if(_ism_debug, "ISM43362Interface: WARNING this firmware version has not been tested !\r\n");
+    debug_if(_ism_debug, "ISM43362Interface: read_version = %lu\r\n", _FwVersion);
+#if TARGET_DISCO_L475VG_IOT01A
+    if (_FwVersion < LATEST_FW_VERSION_NUMBER) {
+        debug_if(_ism_debug, "ISM43362Interface: please update FW\r\n");
     }
```
Comment (LMESTM) [inline ISM43362Interface.cpp]: Update firmware may not be required. But at least inform user that there is more recent one.
"ISM43362Interface: consider upgrading to latest FW " then display version number

File: ISM43362Interface.cpp [hunk 4]
```diff
@@ -18,6 +18,8 @@
 #include "ISM43362Interface.h"
 #include "mbed_debug.h"
 
+#define LATEST_FW_VERSION_NUMBER 3525
```
Comment (LMESTM) [inline ISM43362Interface.cpp]: Please keep the version number in its complete origin format : e.g. C3.5.2.5-STM ...
then use ParseNumber as well when you need to compare with another version

File: ISM43362/ISM43362.cpp (hunk 2 above)
Comment (jeromecoutant) [reply ISM43362/ISM43362.cpp]: @bcostm no, as only the first character is checked. But the i++ is wrong, thx :-)

Comment (jeromecoutant) [reply ISM43362/ISM43362.cpp]: @LMESTM we don't know if it is possible in Inventek driver...
I will keep the code liek this and add a condition to check if result is correct.

File: ISM43362Interface.cpp (hunk 4 above)
Comment (jeromecoutant) [reply ISM43362Interface.cpp]: Done

**Comment by jeromecoutant** [comment]
Rebase and additional patches done

File: ISM43362Interface.cpp [hunk 5]
```diff
@@ -51,23 +47,27 @@ ISM43362Interface::ISM43362Interface(bool debug)
...
-    } else {
-        debug_if(_ism_debug, "ISM43362Interface: WARNING this firmware version has not been tested !\r\n");
+    debug_if(_ism_debug, "ISM43362Interface: read_version = %lu\r\n", _FwVersion);
+    /* FW Revision should be with format "CX.X.X.X" with X as a single digit */
+    if ( (_FwVersion < 1000) || (_FwVersion > 9999) ) {
+        debug_if(_ism_debug, "ISM43362Interface: read_version issue\r\n");
+    }
+
+#if TARGET_DISCO_L475VG_IOT01A
+    if (_FwVersion < ParseNumber(LATEST_FW_VERSION_NUMBER, NULL)) {
```
Comment (adustm) [inline ISM43362Interface.cpp]:
```
[Warning] ISM43362Interface.cpp@59,64: deprecated conversion from string constant to 'char*' [-Wwrite-strings]
[Warning] ISM43362Interface.cpp@59,20: comparison between signed and unsigned integer expressions [-Wsign-compare]
[DEBUG] Return: 0
[DEBUG] Output: C:\MCD8\WIFI\mbed-os\wifi-ism43362\ISM43362Interface.cpp: In constructor 'ISM43362Interface::ISM43362Interface(bool)':
[DEBUG] Output: C:\MCD8\WIFI\mbed-os\wifi-ism43362\ISM43362Interface.cpp:59:64: warning: deprecated conversion from string constant to 'char*' [-Wwrite-strings]
[DEBUG] Output:      if (_FwVersion < ParseNumber(LATEST_FW_VERSION_NUMBER, NULL)) {
[DEBUG] Output:                                                                 ^
[DEBUG] Output: C:\MCD8\WIFI\mbed-os\wifi-ism43362\ISM43362Interface.cpp:59:20: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
[DEBUG] Output:      if (_FwVersion < ParseNumber(LATEST_FW_VERSION_NUMBER, NULL)) {
[DEBUG] Output:                     ^
```
Could you remove this warning please ?

Kind regards

> Kind regards

Cheers

**Pull request merged**